lib/core/test/integration/Test/Integration/Scenario/CLI/Transactions.hs

```diff
@@ -644,6 +651,17 @@ spec = do
             out `shouldBe` ""
             c `shouldBe` ExitFailure 1
 
+    it "TRANS_LIST_01 - Listing transactions for an empty wallet" $ \ctx ->
+        withMaxSuccess 10 $ property $ \mAfter mBefore -> do
```

Comment:
This property is rather slow, so we only check it a few times.

Reply:
Maybe integration tests _layer_ is therefore not the best place for properties? I think that more value we can gain from adding some negative cases here (see my previous comment) and have properties like this on _unit_ level...

Reply:
> Maybe integration tests _layer_ is therefore not the best place for properties? I think that more value we can gain from adding some negative cases here (see my previous comment) and have properties like this on _unit_ level...

Hmm, I wonder if there's a good reason to avoid using properties here?

My intent here was to cover all the cases for `--after` and `--before`. (Both of these parameters can be either present or absent.) The point is that the values can be anything, and the result should still be `[]`.

It seemed sensible to generate these cases rather than write them manually. But perhaps there is a better way?

Reply:
> It seemed sensible to generate these cases rather than write them manually. But perhaps there is a better way?

I agree ^. 

By saying: 
> Maybe integration tests layer is therefore not the best place for properties?

I was reffering to your comment:
>This property is rather slow, so we only check it a few times.

I was thinking that maybe it just takes a lot of time on this _level_ and that's why I though that bringing it to _unit_ level would make it faster...

Reply:
> > This property is rather slow, so we only check it a few times.
> 
> I was thinking that maybe it just takes a lot of time on this _level_ and that's why I though that bringing it to _unit_ level would make it faster...

Good point. I guess it depends on what we deem acceptable. On my system, with 10 cases, it takes around 1-2 seconds to run. Is this fast enough?

Reply:
>On my system, with 10 cases, it takes around 1-2 seconds to run. Is this fast enough?

Yeah, I think that's OK in general. (on the other hand that may be the time of 1000 tests on _unit level_ ;) )

There's also this concept of testing pyramid (https://medium.com/@timothy.cochran/test-pyramid-the-key-to-good-automated-test-strategy-9f3d7e3c02d5), which I'm fond of, which basically says to have the most tests on the lowest level of the system and the least on the highest - end-to-end level. Following this concept we could have say, property doing 1000 examples on _unit test_ level testing a lot of different values and verify that proper error is _thrown_ and have just one integration test with single value (or just a few values) testing if everything works fine _end-to-end_ (e.g. check the error message is presented correctly to the user etc.)

Reply:
> There's also this concept of testing pyramid (https://medium.com/@timothy.cochran/test-pyramid-the-key-to-good-automated-test-strategy-9f3d7e3c02d5), which I'm fond of, which basically says to have the most tests on the lowest level of the system and the least on the highest - end-to-end level.

Thanks for sharing this, the content makes a lot of sense.

For this case though, I'm guessing we really would like to test all the following cases:

1. Both `--after` and `--before` parameters present.
2. Only `--after` present.
3. Only `--before` present.
4. Neither `--after` nor `--before` present.

In addition, to be able to test these parameters with random rather than fixed values, is probably useful.

We could make an argument for writing the above four cases as separate unit tests, but since all cases should produce the same output (the empty list), wouldn't that mean an unnecessary duplication of code? (Also, given that we only generate 10 cases, I'm guessing it wouldn't save us much time.)

Though having said that, I don't have a particularly strong preference in this case! If you have a strong preference, would be happy to rewrite this without properties.

Reply:
We avoid property-based tests in integration tests because:

- We want to keep them predictable. They're already sometimes quite tricky to debug, if we can't reproduce them exactly as we want, it becomes even harder.
- They're slow, so we've to limit their execution to only a small number, and here, we don't have any guarantee that all cases are covered. Ideally, we would use a `checkCoverage` + `cover` to make sure we do cover all the cases we want, but the test would take minutes to run. 
- Instead, when tests a rather similar in their setup, execution and assertion, we use a test matrix to make it clear. See also: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/test/integration/Test/Integration/Scenario/CLI/Addresses.hs#L123-L134

Reply:
@KtorZ @piotr-iohk 

Okay. I've:

- [x] changed the integration test to use a matrix of hand-picked time values (rather than using a property-based test).
- [x] changed the `listTransactionsViaCLI` function to take `Maybe String` for each time value.

Hopefully this should address the concerns raised above, but let me know if you'd like me to change anything about this.

Thanks again for looking at this PR. :)